lnrpc/routerrpc/router_server.go

```diff
@@ -356,20 +360,62 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
 	)
 }
 
-// EstimateRouteFee allows callers to obtain a lower bound w.r.t how much it
-// may cost to send an HTLC to the target end destination.
+// EstimateRouteFee allows callers to obtain an upper bound w.r.t how much it
+// may cost to send an HTLC to the target end destination. This method sends
+// probe payments to the target node, based on target invoice parameters and a
+// random payment hash that makes it impossible for the target to settle the
+// htlc. If provided with a destination key and amount, this method will perform
+// a local graph based fee estimation.
 func (s *Server) EstimateRouteFee(ctx context.Context,
 	req *RouteFeeRequest) (*RouteFeeResponse, error) {
 
-	if len(req.Dest) != 33 {
+	isProbeDestination := len(req.Dest) > 0
+	isProbeInvoice := len(req.PaymentRequest) > 0
+
+	switch {
+	case isProbeDestination && isProbeInvoice:
+		return nil, errors.New("cannot specify both a destination " +
+			"and an invoice")
+
+	case isProbeDestination && len(req.Dest) != 33:
 		return nil, errors.New("invalid length destination key")
+
+	case isProbeDestination && req.AmtSat <= 0:
+		return nil, errors.New("requested probe amount needs to be > 0")
+
+	case isProbeDestination:
+		return s.probeDestination(req.Dest, req.AmtSat)
+
+	case isProbeInvoice:
+		// Consider user provided timeout, otherwise use default.
+		deadLine := time.Now().Add(s.cfg.FeeEstimationTimeout)
+		if req.Timeout > 0 {
+			deadLine = time.Now().Add(
+				time.Duration(req.Timeout) * time.Second,
+			)
+		}
+		probeCtx, cancel := context.WithDeadline(ctx, deadLine)
+		defer cancel()
+
+		return s.probeInvoice(probeCtx, req.PaymentRequest, req.Timeout)
+	}
+
+	return nil, errors.New("specify either a destination or an invoice")
```

Comment:
I'd move this case to the top so we can make progressively more assumptions as the case statements go on.

Reply:
I think this fits into the default case of the switch here.